### PR TITLE
[WIP] Blocking and approximation changes

### PR DESCRIPTION
**Pull request by sergeyf**
Lots here...

1. Removing flake since we have black and it's good enough
2. Bumping version of Python to 3.8.15, same for sklearn and pandas.
3. Changes above triggered more mypy errors hence lots of new "ignore: type"
4. A more slimmed down name tuples file based on ORCID data. [FUNCTIONALITY CHANGE]
5. Found some edge-case issues with emails - stripping periods now. [FUNCTIONALITY CHANGE]
6. Complete subblocking algorithm which uses predict_incremental on the single letter first name signatures internally. [FUNCTIONALITY CHANGE]
7. Subblocking can also be used in `predict_incremental` to batch the inputs. 
8. Subblocking can be used to prevent OOM errors in both `predict` and `predict_incremental` via `batching_threshold` input param.
9. Ability for predict_incremental to first cluster the unassigned signatures together. This makes it slower but improves quality when a lot of the unassigned signatures are actually from the same clusters. [FUNCTIONALITY CHANGE]
10. Added support for partial supervision to predict_incremental. This wasn't needed before as predict_incremental was never called internally, but will be needed for some of the new abilities.

### CODE REVIEW

**Comment by sergeyf** [comment]
Let's schedule time to go through this.

File: s2and/featurizer.py [hunk 1]
```diff
@@ -394,10 +394,10 @@ def _single_pair_featurize(work_input: Tuple[str, str], index: int = -1) -> Tupl
         email_2 = email_2 if "@" in email_2 else email_2 + "@MISSING"
         split_email_1 = email_1.split("@")
         split_email_2 = email_2.split("@")
-        email_prefix_1 = "".join(split_email_1[:-1])
-        email_prefix_2 = "".join(split_email_2[:-1])
-        email_suffix_1 = split_email_1[-1]
-        email_suffix_2 = split_email_2[-1]
+        email_prefix_1 = "".join(split_email_1[:-1]).strip(".")
```
[the remark this answers is not in the document]
Comment (sergeyf) [reply s2and/featurizer.py:397]: It is not. I'll add a lowercase. Thanks.

File: s2and/model.py [hunk 2]
```diff
@@ -457,7 +462,207 @@ def predict(
...
+                block_dict_subblocked_multiple_letter_first_names = block_dict_subblocked_single_letter_first_names
+                block_dict_subblocked_single_letter_first_names = {}
+
+            pred_clusters = {}
+            # ideally we would batch the subblocks for predictions
+            # but it's hard to know how to batch since this can be called
+            # from inside of predict_incremental, which has different OOM behavior.
+            # so just doing it one at a time here
+            if len(block_dict_subblocked_multiple_letter_first_names) > 0:
+                logger.info("Running predict on subblocks with multiple letter first names")
```
[the remark this answers is not in the document]
Comment (sergeyf) [reply s2and/model.py:553]: Will do.

File: s2and/subblocking.py [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+    except:
+        X = X_specter
+
+    # how many subblocks do we want given this data and target subblock size?
+    # should be at least 2 if we end up here otherwise there is no point
+    num_desired_subblocks = int(np.ceil(len(signature_ids) / target_subblock_size))
+
+    # this can fail when X are all zeros
+    try:
+        g = genieclust.Genie(n_clusters=num_desired_subblocks, gini_threshold=0.01, exact=False)
```
[the remark this answers is not in the document]
Comment (sergeyf) [reply s2and/subblocking.py:74]: I messed around with this extensively and anything large just made one giant cluster and a lot of tiny ones. It's essentially useless at that point.

**Review by Whattabatt — APPROVED** [review]
Looks good to me, though of course we'll have to do more investigation / engineering discussion before working these changes into the production setup

**Pull request merged**